maskit/datasets/mnist.py

```diff
@@ -48,51 +30,95 @@ def apply_PCA(wires: int, x_train: np.ndarray):
     return pca
 
 
+def downscale(x_data, y_data, size) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    This function does several things including the reduction of image size,
+    conversion to binary as well as removal of duplicates.
+
+    :param x_data: Input data
+    :param y_data: Target data
+    :param size: Maximum number of data to prepare
+    """
+    data_index: List[bool] = np.zeros(len(y_data), dtype=bool)
+    result_x: List[np.ndarray] = []
+    result_x_set = set()
+    for index, image in enumerate(x_data):
+        if len(result_x) >= size:
+            break
+        reduced_image = reduce_image(image)
+        previous_len = len(result_x_set)
+        result_x_set.add(str(reduced_image))
+        if len(result_x_set) <= previous_len:  # next if contradicting
+            continue
+        result_x.append(reduced_image)
+        data_index[index] = True
+    return (np.array(result_x), y_data[data_index])
+
+
+def prepare_data(
+    x_data: np.ndarray, y_data: np.ndarray, size, classes
+) -> Tuple[np.ndarray, np.ndarray]:
+    data_index = np.isin(y_data, classes)
+    x_data, y_data = x_data[data_index], y_data[data_index]
+    x_data, y_data = downscale(x_data, y_data, size)
+    y_data = np.array([convert_label(y, classes) for y in y_data])
+    try:
+        x_data = np.reshape(
+            x_data, (x_data.shape[0], x_data.shape[1] * x_data.shape[2])
+        )
+    except IndexError:
+        pass
+    return (x_data, y_data)
+
+
 def mnist(
-    wires=4, classes=(6, 9), train_size=100, test_size=50, shuffle=True
+    wires: int = 4,
+    classes=(6, 9),
+    train_size: int = 100,
+    validation_size: int = 0,
+    test_size: int = 50,
+    shuffle: bool = True,
 ) -> DataSet:
     (x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
-    train_size = min(train_size, MAX_TRAIN_SAMPLES)
+    train_size = min(train_size, MAX_TRAIN_SAMPLES - validation_size)
+    validation_size = min(validation_size, MAX_TRAIN_SAMPLES - train_size)
     test_size = min(test_size, MAX_TEST_SAMPLES)
 
-    x_train, y_train = zip(*((x, y) for x, y in zip(x_train, y_train) if y in classes))
-    x_test, y_test = zip(*((x, y) for x, y in zip(x_test, y_test) if y in classes))
-
-    x_train = [reduce_image(x) for x in x_train]
-    x_test = [reduce_image(x) for x in x_test]
-
-    x_train, y_train = remove_contradicting(x_train, y_train)
-    x_test, y_test = remove_contradicting(x_test, y_test)
-
-    x_train, y_train = x_train[:train_size], y_train[:train_size]
-    x_test, y_test = x_test[:test_size], y_test[:test_size]
-
-    x_train = convert_to_binary(x_train)
-    x_test = convert_to_binary(x_test)
-    y_train = [convert_label(y, classes) for y in y_train]
-    y_test = [convert_label(y, classes) for y in y_test]
+    if shuffle:
+        x_train, y_train = skl_shuffle(x_train, y_train)
+
+    # split validation set from train set
+    split_point = len(x_train) - (5000 if validation_size > 0 else 0)
```

Comment:
Is there a reason why this is always 5000 and not the validation_size?

Reply:
Yes, indeed! 🤓
I borrowed this number based on a Tensorflow's MNIST tutorial that used 5,000 images as a validation set. However, there is also a paper on *What is the Best Multi-Stage Architecture for Object Recognition?*, in which they use a set of 10,000 validation images. So I am glad to change this if you have a specific opinion on this.

Reply:
You could do the splitting of training and validation set after calling ```prepare_data()``` with ```train_size + validation_size``` instead of calling it for each one individually.

But then you have to decide what to do if there are not enough data points for both of them. Reduce the size of the training set or of the validation set?

Reply:
I just created a ticket out of this discussion (see #63) so we can postpone any decisions.